pennylane/measurements/purity.py

```diff
@@ -0,0 +1,113 @@
+
+    .. math::
+        \gamma = \text{Tr}(\rho^2)
+
+    where :math:`\rho` is the density matrix. The purity of a normalized quantum state satisfies
+    :math:`\frac{1}{d} \leq \gamma \leq 1`, where :math:`d` is the dimension of the Hilbert space.
+    A pure state has a purity of 1.
+
+    Args:
+        wires (Sequence[int] or int): The wires of the subsystem
+
+    Returns:
+        PurityMP: Measurement process instance
+
+    **Example**
+
+    .. code-block:: python3
+
+        dev = qml.device("default.mixed", wires=2)
+
+        @qml.qnode(dev)
+        def circuit_purity(p):
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.BitFlip(p, wires=0)
+            qml.BitFlip(p, wires=1)
+            return qml.purity(wires=[0,1])
+
+    >>> circuit_purity(0.1)
+    array(0.7048)
+
+    .. seealso:: :func:`pennylane.qinfo.transforms.purity` and :func:`pennylane.math.purity`
+    """
+    wires = Wires(wires)
+    return PurityMP(wires=wires)
+
+
+class PurityMP(StateMeasurement):
+    """Measurement process that computes the purity of the system prior to measurement.
+
+    Please refer to :func:`purity` for detailed documentation.
+
+    Args:
+        obs (.Observable): The observable that is to be measured as part of the
+            measurement process. Not all measurement processes require observables (for
+            example ``Probability``); this argument is optional.
+        wires (.Wires): The wires the measurement process applies to.
+            This can only be specified if an observable was not provided.
+        eigvals (array): A flat array representing the eigenvalues of the measurement.
+            This can only be specified if an observable was not provided.
+        id (str): custom label given to a measurement instance, can be useful for some
+            applications where the instance has to be identified
+    """
+
+    @property
+    def return_type(self):
+        return Purity
+
+    @property
+    def numeric_type(self):
+        return float
+
+    def shape(self, device=None):
+        if qml.active_return():
+            return self._shape_new(device)
+        if device is None or device.shot_vector is None:
+            return (1,)
+        num_shot_elements = sum(s.copies for s in device.shot_vector)
+        return (num_shot_elements,)
+
+    def _shape_new(self, device=None):
+        if device is None or device.shot_vector is None:
+            return ()
+        num_shot_elements = sum(s.copies for s in device.shot_vector)
+        return tuple(() for _ in range(num_shot_elements))
+
+    def process_state(self, state: Sequence[complex], wire_order: Wires):
+        if wire_order.labels != tuple(range(len(wire_order))):
+            raise qml.QuantumFunctionError(
+                "Returning the purity is not supported when using custom wire labels"
+            )
+        return qml.math.purity(state, indices=self.wires, c_dtype=state.dtype)
```

Comment:
You can easily compute the indices of the measurement by creating a `wire_map` with the `wire_order` argument:
```suggestion
    def process_state(self, state: Sequence[complex], wire_order: Wires):
        wire_map = dict(zip(wire_order, list(range(len(wire_order)))))
        indices = [wire_map[w] for w in self.wires]
        return qml.math.purity(state, indices=indices, c_dtype=state.dtype)
```